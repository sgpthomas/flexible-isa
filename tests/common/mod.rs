use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
