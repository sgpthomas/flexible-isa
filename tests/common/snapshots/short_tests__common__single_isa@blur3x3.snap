---
source: tests/common/mod.rs
description: generate isa for blur3x3
---
[
    (
        0,
        1,
        "(neg (cast<int32> ?x0))",
    ),
    (
        4,
        1,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 2))))) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> (neg (cast<int32> 64))))",
    ),
    (
        7,
        1,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> 64))))) (cast<int32> ?x2))",
    ),
    (
        19,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> (\"%\" (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (call[max] (cast<int32> ?#2) (cast<int32> 64))))) (cast<int32> (neg (cast<int32> 64))))",
    ),
    (
        22,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 2))))) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        27,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> 64))))",
    ),
    (
        30,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        42,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        47,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 128))))",
    ),
    (
        64,
        1,
        "(- ?#0 ?#1)",
    ),
    (
        75,
        1,
        "(* (cast<int32> ?x0) (cast<int32> (call[max] (cast<int32> ?#1) (cast<int32> 64))))",
    ),
    (
        83,
        1,
        "(/ (cast<uint16x64> (+ (cast<uint16x64> (get ?#9 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 64))))) (cast<uint16x64> (+ (cast<uint16x64> (get ?#10 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x3) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 64))))) (cast<uint16x64> (get ?#11 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x4) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 64))))))))) (cast<uint16x64> (call[x64] (cast<uint16> 3))))",
    ),
    (
        85,
        1,
        "(/ ?x0 (cast<uint16x64> (call[x64] (cast<uint16> 3))))",
    ),
    (
        87,
        1,
        "(\"%\" (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> 4))",
    ),
    (
        88,
        1,
        "(\"%\" (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 3))) (cast<int32> 4))",
    ),
    (
        89,
        1,
        "(\"%\" (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 4))",
    ),
    (
        91,
        1,
        "(\"%\" (cast<int32> ?x0) (cast<int32> 4))",
    ),
    (
        100,
        1,
        "(<= (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        103,
        1,
        "(== ?#0 ?#1)",
    ),
    (
        108,
        1,
        "(&& (cast<bool> (<= (cast<int32> ?x0) (cast<int32> ?x1))) (cast<bool> ?x2))",
    ),
    (
        110,
        1,
        "(&& (cast<bool> ?x0) (cast<bool> (<= (cast<int32> ?x1) (cast<int32> ?x2))))",
    ),
    (
        119,
        1,
        "(get ?#1 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (\"%\" (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (call[max] (cast<int32> ?#2) (cast<int32> 64))))) (cast<int32> (neg (cast<int32> 64))))) (cast<int32> 1) (cast<int32> 64))))",
    ),
    (
        36,
        2,
        "(+ (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (* (cast<int32> ?#1) (cast<int32> 128))) (cast<int32> 384))) (cast<int32> ?#2))) (cast<int32> ?x0))",
    ),
    (
        41,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        53,
        2,
        "(+ (cast<uint16x64> (get ?#2 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> (\"%\" (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> 1))) (cast<int32> (call[max] (cast<int32> ?#3) (cast<int32> 64))))) (cast<int32> (neg (cast<int32> 64))))) (cast<int32> 1) (cast<int32> 64))))) (cast<uint16x64> ?x1))",
    ),
    (
        67,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 2))))) (cast<int32> ?x1))",
    ),
    (
        73,
        2,
        "(* (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        74,
        2,
        "(* (cast<int32> ?x0) (cast<int32> 64))",
    ),
    (
        76,
        2,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        79,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        84,
        2,
        "(/ (cast<uint8x64> (+ (cast<uint8x64> (get ?#6 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?x1) (cast<int32> (* (cast<int32> ?#7) (cast<int32> 2))))))) (cast<int32> (neg (cast<int32> 63))))) (cast<int32> 1) (cast<int32> 64))))) (cast<uint8x64> (+ (cast<uint8x64> (get ?#8 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?x1) (cast<int32> (* (cast<int32> ?#9) (cast<int32> 2))))))) (cast<int32> (neg (cast<int32> 62))))) (cast<int32> 1) (cast<int32> 64))))) (cast<uint8x64> (get ?#10 (cast<int32x64> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (+ (cast<int32> ?x1) (cast<int32> (* (cast<int32> ?#11) (cast<int32> 2))))))) (cast<int32> (neg (cast<int32> 64))))) (cast<int32> 1) (cast<int32> 64))))))))) (cast<uint16x64> (call[x64] (cast<uint16> 3))))",
    ),
    (
        80,
        3,
        "(/ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> 64))",
    ),
    (
        81,
        3,
        "(/ (cast<int32> ?x0) (cast<int32> 64))",
    ),
    (
        104,
        3,
        "(!= (cast<int32> (\"%\" (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> 64))) (cast<int32> 0))",
    ),
    (
        105,
        3,
        "(!= (cast<int32> (\"%\" (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> 0))",
    ),
    (
        107,
        3,
        "(&& (cast<bool> (<= (cast<int32> (+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> (/ (cast<int32> (+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> 63))) (cast<int32> 64))) (cast<int32> 64))) (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> 64))))) (cast<int32> (neg (cast<int32> 64))))) (cast<int32> ?x0))) (cast<int32> (+ (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> (+ (cast<int32> (* (cast<int32> (/ (cast<int32> (+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> 64))) (cast<int32> 63))) (cast<int32> 64))) (cast<int32> 64))) (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> 64))))))) (cast<int32> (neg (cast<int32> 63))))))) (cast<int32> 1))) ?x1)) (cast<bool> (<= (cast<int32> 64) (cast<int32> ?x0))))",
    ),
    (
        52,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
]
